### ISSUE TITLE
strange mismatch behavior UDP OUT OF ORDER 

### ISSUE THREAD

**Issue by bbordereau** [esnet/iperf#457]
Hello, i have got a strange behavior
On productio network , i decide to test bandwidth and error. So i perform an end to end between a client and a server in DC.
When i perform test the iperf3 with the server as server ( iperf3 -s ) and client as client in UDP , with 10Mb/s everything is ok  but 100Mb/s is being worse and i see in the server side, a lot of messages as OUT OF ORDER 
I do not see errors or drops on my network 
When i reverse side, the lost is worts more than 72%
If i did the test with TCP , everything looks fine 
is anyone has got an idea , i would like to be sure that my issue is not iperf3 itself ?
the command for client i used : iperf3 -c <@ip server) - b 100M -u 
any help will be great ?
what does it mean out of order in iperf3, and what are the possible root cause 
Best regards
Bert

**Comment by bmah888** [comment]
This means that the iperf3 server is receiving the UDP test packets in a different order than they were sent from the iperf3 client.  There could be a variety of explanations for this phenomenon, ranging from packets taking different paths through the network to bugs in network drivers.

More information would be required for someone to help you diagnose the problem.  If you take a packet trace (say using tcpdump) at the destination, you could actually decode the iperf3 sequence numbers in the packets and see if the packets had arrived intact, out of order, etc.

**Comment by bltierney** [comment]
might there be a bonded link or bonded NIC? That often leads to out of order packets.

Also, please see if 3.1.6 fixes it.

**Comment by AndrewJDR** [comment]
@bmah888 :

Given that UDP as a protocol makes no guarantee that packets will arrive in the same order they're sent, why do out of order packets get counted as a "lost" packets in the stats output?

For example, in the UDP test output below using the latest f1415a0d98b commit of iperf on both ends and using the command `iperf3 -R -u --get-server-output -b 4M -c <my_server_hostname>`, there are several "OUT OF ORDER" messages, then immediately afterward, those out of order packets are reported as lost packets (e.g. `2/342`). Why are they counted as lost? This is UDP, so out-of-order packets are something that can happen under totally normal circumstances.
```
Connecting to host xxx, port 5201
Reverse mode, remote host xxx is sending
[  4] local 192.168.1.110 port 56316 connected to yyy port 5201
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-1.00   sec   510 KBytes  4.18 Mbits/sec  1.091 ms  0/358 (0%)
iperf3: OUT OF ORDER - incoming packet = 537 and received packet = 538 AND SP = 4
[  4]   1.00-2.00   sec   489 KBytes  4.01 Mbits/sec  1.111 ms  1/343 (0.29%)
[  4]   2.00-3.00   sec   488 KBytes  3.99 Mbits/sec  0.439 ms  0/342 (0%)
[  4]   3.00-4.00   sec   489 KBytes  4.01 Mbits/sec  1.150 ms  0/343 (0%)
iperf3: OUT OF ORDER - incoming packet = 1562 and received packet = 1563 AND SP = 4
iperf3: OUT OF ORDER - incoming packet = 1581 and received packet = 1582 AND SP = 4
[  4]   4.00-5.00   sec   488 KBytes  3.99 Mbits/sec  1.397 ms  2/342 (0.58%)
iperf3: OUT OF ORDER - incoming packet = 2002 and received packet = 2004 AND SP = 4
[  4]   5.00-6.00   sec   488 KBytes  4.00 Mbits/sec  1.180 ms  1/342 (0.29%)
[  4]   6.00-7.00   sec   489 KBytes  4.01 Mbits/sec  0.753 ms  0/343 (0%)
[  4]   7.00-8.00   sec   489 KBytes  4.01 Mbits/sec  0.989 ms  0/343 (0%)
[  4]   8.00-9.00   sec   488 KBytes  3.99 Mbits/sec  2.294 ms  0/342 (0%)
[  4]   9.00-10.00  sec   486 KBytes  3.98 Mbits/sec  1.401 ms  0/341 (0%)
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-10.06  sec  4.80 MBytes  4.00 Mbits/sec  0.000 ms  0/3445 (0%)  sender
[SUM]  0.0-10.1 sec  4 datagrams received out-of-order
[  4]   0.00-10.00  sec  4.80 MBytes  4.02 Mbits/sec  1.361 ms  4/3445 (0.12%)  receiver

Server output:
-----------------------------------------------------------
Server listening on 5201
-----------------------------------------------------------
Accepted connection from zzz, port 35815
[  6] local yyy port 5201 connected to zzz port 39455
[ ID] Interval           Transfer     Bandwidth       Total Datagrams
[  6]   0.00-1.00   sec   489 KBytes  4.01 Mbits/sec  343
[  6]   1.00-2.00   sec   488 KBytes  3.99 Mbits/sec  342
[  6]   2.00-3.00   sec   489 KBytes  4.01 Mbits/sec  343
[  6]   3.00-4.00   sec   488 KBytes  3.99 Mbits/sec  342
[  6]   4.00-5.00   sec   489 KBytes  4.01 Mbits/sec  343
[  6]   5.00-6.00   sec   488 KBytes  3.99 Mbits/sec  342
[  6]   6.00-7.00   sec   488 KBytes  3.99 Mbits/sec  342
[  6]   7.00-8.00   sec   489 KBytes  4.01 Mbits/sec  343
[  6]   8.00-9.00   sec   488 KBytes  3.99 Mbits/sec  342
[  6]   9.00-10.00  sec   489 KBytes  4.01 Mbits/sec  343
[  6]  10.00-10.06  sec  28.5 KBytes  4.01 Mbits/sec  20
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  6]   0.00-10.06  sec  4.80 MBytes  4.00 Mbits/sec  0.000 ms  0/3445 (0%)  sender


iperf Done.
```

**Comment by bmah888** [comment]
@AndrewJDR:  I'm not sure, to be perfectly honest, why it's counting errors in the way it is.  The code in question was written before I started working on this project, or even before I joined ESnet.

My guess (and that's all it is) is that if two packets were interchanged, there then appears to be a gap in received sequence numbers, which gets (wrongly) counted as a loss.  In other words, if the packets were sent in the order ABCD but received in the order ACBD, then iperf3 might be thinking there was a packet lost between B and D.  That's not what happened of course.  I don't have any deeper insights at this point.

**Comment by bltierney** [comment]
I think ideally iperf3 would be smart enough to keep track of loss vs out-of-order separately. Clearly out-of-order packets should not show up on the 'loss' counter.

**Comment by bmah888** [comment]
The code in question is in `iperf_udp_recv`, an excerpt is below:
```
    /* Out of order packets */
    if (pcount >= sp->packet_count + 1) {
        if (pcount > sp->packet_count + 1) {
            sp->cnt_error += (pcount - 1) - sp->packet_count;
        }
        sp->packet_count = pcount;
    } else {
        sp->outoforder_packets++;
	iperf_err(sp->test, "OUT OF ORDER - incoming packet = %zu and received packet = %d AND SP = %d", pcount, sp->packet_count, sp->socket);
    }
```
`pcount` is the sequence number read from an arriving UDP packet.  `sp->packet_count` is the highest sequence number seen so far.  Yes, it appears to be misnamed, this value might not have any relation to the total number of packets received on the stream.  Basically the first conditional checks to see if arriving packet makes the sequence numbers go forwards or backwards.  If forwards, then we count gaps (if any) as lost packets by adding the size of the gap to `sp->cnt_error`.  Then we set the next expected sequence number.

If the packet sequence numbers went backwards, we increment the out-of-order counter `sp->outoforder_packets` and output a slightly confusing error message with the actual and expected sequence numbers and the file descriptor.  The error message goes to `stderr`, which I believe to be a mistake (note that we don't log individual packet losses)...it might be better to hide that error message behind the debug or verbose flags.

Let's suppose a sequence of packets is sent with sequence numbers (1, 2, 3, 4) but somehow received in the sequence (1, 3, 2, 4).  Packet 1 would get processed without incident.  The reception of packet 3 would then count as a loss (because iperf3 would then think that packet 2 had been dropped), but the receiver would then be expecting packet 4 next.  Packet 2 actually gets received, and this increments the out-of-order counter (but packet 4 remains the next expected packet).  Finally packet 4 arrives at the receiver, which is the expected value.  This sequence of packets would count as one out-of-order packet and one loss, even though in reality there were no actual losses.

I postulate that it would be (more) correct that for every out-of-order packet we receive, we can decrement the number of losses (that number should always be non-zero, because we have to have a gap in sequence numbers to have an out-of-order packet, assuming no duplication of packets).  I am sure that there are some weird edge cases with duplicated packets or combinations of loss and reordering that will make this fall over, but as far as I can tell the only way to accurately count loss in scenarios like this is to keep a bitmap of all sequence numbers and tick off the sequence numbers that we've actually seen.  That feels rather cumbersome.

So at this point I'm considering making that slight tweak to the out-of-order processing, changing the now-error message, and adding some internal documentation explaining what the code is doing.  I don't have an easy way to test this however...I'm up for suggestions as to how to test this in a somewhat controlled fashion.

**Comment by AndrewJDR** [comment]
@bmah888 If you want to test out-of-order packets and/or packet loss, you can use the 'tc' tool on linux:
https://wiki.linuxfoundation.org/networking/netem#packet-re-ordering

for e.g.:
```
# First add a qdisc:
tc qdisc add dev eth0 root netem delay 1ms
# For simulating packet loss:
tc qdisc change dev eth0 root netem loss 0.1%
# Or for simulating re-ordering:
tc qdisc change dev eth0 root netem gap 5 delay 10ms
# To switch between loss/re-ordering, first run:
tc qdisc del dev eth0 root
# Then run "tc qdisc add" and the appropriate "tc qdisc change" command above...
```

(There are probably more concise ways of using these commands, but this is the method I'm accustomed to.)

**Comment by bmah888** [comment]
@AndrewJDR:  Thanks for the info, that was a great help to get me started.  On the CentOS 7 VM I'm using, I need to do something slightly different, like this:
```
tc qdisc add dev lo root netem delay 1ms
tc qdisc change dev lo root netem reorder 100 gap 5 delay 10ms
```
Now testing with this server:
```
bmah-centos7:iperf% src/iperf3 --server
```
And this client:
```
bmah-centos7:iperf% src/iperf3 --client localhost --udp -b 15m
```
I needed to increase the UDP sending bitrate on the client, because with the default, it seems like the packets aren't spaced closely enough for the for the netem reordering to have any effect.  15Mbps seems to be fast enough that we see some amount of reordering but we're not overwhelmed by the output.

**Comment by AndrewJDR** [comment]
@bmah888 Thanks for this... I will try this out next time I'm running some iperf tests, and let you know how it works.

**Comment by bmah888** [comment]
I merged that pull request (#589) to master after some testing, on the grounds that the prior behavior was fairly broken, and not likely to get any worse even if my so-called solution doesn't work.  I don't think it's completely correct in all cases BTW, but it counts correctly (or close to correctly) in the cases I've been testing.  Closing this issue for now...it can be re-opened if needed.
